test/testdata/infer/product.rb

```diff
@@ -0,0 +1,28 @@
+# typed: true
+
+foo = T.let(
+  ["a", "b", "c"],
+  T::Array[String]
+)
+
+bar = T.let(
+  [1, 2, 3],
+  T::Array[Integer]
+)
+
+mixed_type = T.let(
+  [1, "alpha", 0.0],
+  T::Array[T.any(String, Integer, Float)]
+)
+
+empty = T.let(
+  [],
+  T::Array[T.untyped]
+)
+
+
+T.assert_type!(empty.product, T::Array[T.untyped])
+T.assert_type!(foo.product, T::Array[[String]])
+T.assert_type!(foo.product(mixed_type), T::Array[[String, T.any(String, Integer, Float)]])
+T.assert_type!(foo.product(bar), T::Array[[String, Integer]])
+T.assert_type!(foo.product(bar, foo), T::Array[[String, Integer, String]])
```

Comment:
There are no tests here that exercise the TupleType cases. These all operate on args and receivers that are Arrays.

I'd also like to see a test that calls product with like 3 or 4 args.

Reply:
And also a test that calls `product` with something that is not an array. You can expect type errors on certain lines in tests by putting `# error: (the message)` at the end of a line that should have an error.

Reply:
Added multiple args, TupleType, as well as error cases